File: duck.py
"""
Duck typing example in python
"""

import math

def getMass(p):
	"""
	This function requires that p implement getEnergy and getMomentum
	methods, but does not care about their formal types
	"""
	print 'Calling getMass for type',type(p)
	return math.sqrt(p.getEnergy()**2 - p.getMomentum()**2)

class Electron(object):
	def __init__(self,p):
		self.p = p
		self.m = 0.511
		self.E = math.sqrt(p**2+self.m**2)
	def getEnergy(self):
		return self.E
	def getMomentum(self):
		return self.p

class SpaceShuttle(object):
	def __init__(self,fuel):
		self.fuel = fuel
		self.p = 10.
		self.m = 10000.
		self.E = math.sqrt(self.p**2+self.m**2)
	def getEnergy(self):
		return self.E
	def getMomentum(self):
		return self.p
	def getFuel(self):
		return self.fuel

electron = Electron(1)
print getMass(electron)

shuttle = SpaceShuttle(500)
print getMass(shuttle)

File: duck.cc
// Duck typing example in C++
// Compile and test with:
// gcc duck.cc
// .a/out

#include <cmath>
#include <iostream>

template <typename P>
double getMass(P const &p) {
	return std::sqrt(p.getEnergy()*p.getEnergy() - p.getMomentum()*p.getMomentum());
}

class Electron {
public:
	Electron(double p) : _p(p), _m(0.511) {
		_E = std::sqrt(p*p + _m*_m);
	}
	double getEnergy() const { return _E; }
	double getMomentum() const { return _p; }
private:
	double _m,_E,_p;
};

class SpaceShuttle {
public:
	SpaceShuttle(double fuel) : _fuel(fuel), _p(10), _m(10000) {
		_E = std::sqrt(_p*_p + _m*_m);
	}
	double getEnergy() const { return _E; }
	double getMomentum() const { return _p; }
	double getFuel() const { return _fuel; }
private:
	double _m,_E,_p,_fuel;
};

main() {
	Electron electron(1);
	std::cout << "Electron mass is " << getMass(electron) << std::endl;
	SpaceShuttle shuttle(1);
	std::cout << "Electron mass is " << getMass(shuttle) << std::endl;
}
